The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3824940560/jobs/6507521868) [(plain)](https://github.com/rust-lang/rust/commit/42ab7a46ea2465e3a43a485ebbbfd98ced2a9ac4/checks/10401199132/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   |
note: associated function defined here
  --> src/tools/jsondoclint/src/validator.rs:44:12
   |
44 |     pub fn new(krate: &'a Crate, krate_json: Value) -> Self {
help: provide the argument
   |
   |
9  |     let mut validator = Validator::new(krate, /* serde_json::Value */);

error[E0308]: mismatched types
  --> src/tools/jsondoclint/src/validator/tests.rs:49:31
   |
   |
49 |     check(&k, &[Error { kind: ErrorKind::NotFound, id: id("1") }]);
   |                               ^^^^^^^^^^^^^^^^^^^ expected enum `ErrorKind`, found fn item
  ::: src/tools/jsondoclint/src/main.rs:22:5
   |
   |
22 |     NotFound(Vec<json_find::Selector>),
   |     -------- fn(Vec<Vec<json_find::SelectorPart>>) -> ErrorKind {ErrorKind::NotFound} defined here
   = note: expected enum `ErrorKind`
   = note: expected enum `ErrorKind`
           found fn item `fn(Vec<Vec<json_find::SelectorPart>>) -> ErrorKind {ErrorKind::NotFound}`
help: use parentheses to construct this tuple variant
   |
49 |     check(&k, &[Error { kind: ErrorKind::NotFound(/* Vec<Vec<json_find::SelectorPart>> */), id: id("1") }]);

Some errors have detailed explanations: E0061, E0308.
For more information about an error, try `rustc --explain E0061`.
error: could not compile `jsondoclint` due to 2 previous errors
```

</details>

--- a/src/tools/jsondoclint/Cargo.toml
+++ b/src/tools/jsondoclint/Cargo.toml
@@ -7,6 +7,7 @@ edition = "2021"
 
 [dependencies]
 anyhow = "1.0.62"
+clap = { version = "4.0.15", features = ["derive"] }
 fs-err = "2.8.1"
 rustdoc-json-types = { version = "0.1.0", path = "../../rustdoc-json-types" }
 serde_json = "1.0.85"

--- a/src/tools/jsondoclint/Cargo.toml
+++ b/src/tools/jsondoclint/Cargo.toml
@@ -10,4 +10,5 @@ anyhow = "1.0.62"
 clap = { version = "4.0.15", features = ["derive"] }
 fs-err = "2.8.1"
 rustdoc-json-types = { version = "0.1.0", path = "../../rustdoc-json-types" }
+serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0.85"

--- a/src/tools/jsondoclint/src/json_find.rs
+++ b/src/tools/jsondoclint/src/json_find.rs
@@ -1,8 +1,9 @@
 use std::fmt::Write;
 
+use serde::Serialize;
 use serde_json::Value;
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
 pub enum SelectorPart {
     Field(String),
     Index(usize),

--- a/src/tools/jsondoclint/src/main.rs
+++ b/src/tools/jsondoclint/src/main.rs
@@ -1,6 +1,5 @@
-use std::env;
-
-use anyhow::{anyhow, bail, Result};
+use anyhow::{bail, Result};
+use clap::Parser;
 use fs_err as fs;
 use rustdoc_json_types::{Crate, Id, FORMAT_VERSION};
 use serde_json::Value;
@@ -21,8 +20,15 @@ enum ErrorKind {
     Custom(String),
 }
 
+#[derive(Parser)]
+struct Cli {
+    /// The path to the json file to be linted
+    path: String,
+}
+
 fn main() -> Result<()> {
-    let path = env::args().nth(1).ok_or_else(|| anyhow!("no path given"))?;
+    let Cli { path } = Cli::parse();
+
     let contents = fs::read_to_string(&path)?;
     let krate: Crate = serde_json::from_str(&contents)?;
     assert_eq!(krate.format_version, FORMAT_VERSION);

--- a/src/tools/jsondoclint/src/main.rs
+++ b/src/tools/jsondoclint/src/main.rs
@@ -1,23 +1,32 @@
+use std::io::{BufWriter, Write};
+
 use anyhow::{bail, Result};
 use clap::Parser;
 use fs_err as fs;
 use rustdoc_json_types::{Crate, Id, FORMAT_VERSION};
+use serde::Serialize;
 use serde_json::Value;
 
 pub(crate) mod item_kind;
 mod json_find;
 mod validator;
 
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug, PartialEq, Eq, Serialize, Clone)]
 struct Error {
     kind: ErrorKind,
     id: Id,
 }
 
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug, PartialEq, Eq, Serialize, Clone)]
 enum ErrorKind {
     NotFound(Vec<json_find::Selector>),
     Custom(String),
+}
+
+#[derive(Debug, Serialize)]
+struct JsonOutput {
+    path: String,
+    errors: Vec<Error>,
 }
 
 #[derive(Parser)]
@@ -28,10 +37,13 @@ struct Cli {
     /// Show verbose output
     #[arg(long)]
     verbose: bool,
+
+    #[arg(long)]
+    json_output: Option<String>,
 }
 
 fn main() -> Result<()> {
-    let Cli { path, verbose } = Cli::parse();
+    let Cli { path, verbose, json_output } = Cli::parse();
 
     let contents = fs::read_to_string(&path)?;
     let krate: Crate = serde_json::from_str(&contents)?;
@@ -41,6 +53,13 @@ fn main() -> Result<()> {
 
     let mut validator = validator::Validator::new(&krate, krate_json);
     validator.check_crate();
+
+    if let Some(json_output) = json_output {
+        let output = JsonOutput { path: path.clone(), errors: validator.errs.clone() };
+        let mut f = BufWriter::new(fs::File::create(json_output)?);
+        serde_json::to_writer(&mut f, &output)?;
+        f.flush()?;
+    }
 
     if !validator.errs.is_empty() {
         for err in validator.errs {

--- a/src/tools/jsondoclint/src/main.rs
+++ b/src/tools/jsondoclint/src/main.rs
@@ -24,10 +24,14 @@ enum ErrorKind {
 struct Cli {
     /// The path to the json file to be linted
     path: String,
+
+    /// Show verbose output
+    #[arg(long)]
+    verbose: bool,
 }
 
 fn main() -> Result<()> {
-    let Cli { path } = Cli::parse();
+    let Cli { path, verbose } = Cli::parse();
 
     let contents = fs::read_to_string(&path)?;
     let krate: Crate = serde_json::from_str(&contents)?;
@@ -53,11 +57,27 @@ fn main() -> Result<()> {
                             err.id.0,
                             json_find::to_jsonpath(&sel)
                         ),
-                        [sel, ..] => eprintln!(
-                            "{} not in index or paths, but refered to at '{}' and more",
-                            err.id.0,
-                            json_find::to_jsonpath(&sel)
-                        ),
+                        [sel, ..] => {
+                            if verbose {
+                                let sels = sels
+                                    .iter()
+                                    .map(json_find::to_jsonpath)
+                                    .map(|i| format!("'{i}'"))
+                                    .collect::<Vec<_>>()
+                                    .join(", ");
+                                eprintln!(
+                                    "{} not in index or paths, but refered to at {sels}",
+                                    err.id.0
+                                );
+                            } else {
+                                eprintln!(
+                                    "{} not in index or paths, but refered to at '{}' and {} more",
+                                    err.id.0,
+                                    json_find::to_jsonpath(&sel),
+                                    sels.len() - 1,
+                                )
+                            }
+                        }
                     }
                 }
                 ErrorKind::Custom(msg) => eprintln!("{}: {}", err.id.0, msg),

